src/Amalgam/evaluablenode/EvaluableNodeTreeFunctions.h

```diff
@@ -35,156 +35,282 @@ class CustomEvaluableNodeComparator
-//if id_path specifies the entity in from_entity, then it returns a reference to it
-//if dest_sid_with_reference is not null, it will assume it is a location to store a string id with reference
-// that must be managed by caller
-template<typename EntityReferenceType, typename ContainerEntityReferenceType = EntityReadReference>
-std::pair<EntityReferenceType, ContainerEntityReferenceType>
-TraverseToExistingEntityReferenceAndContainerViaEvaluableNodeIDPath(
-	Entity *from_entity, EvaluableNode *id_path,
-	StringInternPool::StringID *dest_sid_with_reference = nullptr)
+class EvaluableNodeIDPathTraverser
 {
-	//if the destination sid is requested, initialize it
-	if(dest_sid_with_reference != nullptr)
-		*dest_sid_with_reference = string_intern_pool.NOT_A_STRING_ID;
+public:
+	inline EvaluableNodeIDPathTraverser()
+		: idPath(nullptr), idPathEntries(nullptr), curIndex(0)
+	{	}
 
-	//string id to use throughout various flows
-	StringInternPool::StringID sid = string_intern_pool.NOT_A_STRING_ID;
+	//calls AnalyzeIDPath with the same parameters
+	inline EvaluableNodeIDPathTraverser(EvaluableNode *id_path, bool has_destination_id)
+	{
+		AnalyzeIDPath(id_path, has_destination_id);
+	}
 
-	if(from_entity == nullptr)
-		return std::make_pair(EntityReferenceType(nullptr), ContainerEntityReferenceType(nullptr));
+	//populates attribused based on the id_path
+	//if has_destination_id, then it will leave one id at the end for the destination
+	void AnalyzeIDPath(EvaluableNode *id_path, bool has_destination_id)
+	{
+		idPath = id_path;
+		idPathEntries = &idPath->GetOrderedChildNodes();
+		curIndex = 0;
+		containerIdIndex = 0;
+		entityIdIndex = 0;
+		lastIdIndex = 0;
+
+		//size of the entity list excluding trailing nulls
+		size_t non_null_size = idPathEntries->size();
+		while(non_null_size > 0 && EvaluableNode::IsNull((*idPathEntries)[non_null_size - 1]))
+			non_null_size--;
+
+		//if no entities, nothing to traverse
+		if(non_null_size == 0)
+		{
+			idPathEntries = nullptr;
+			return;
+		}
 
-	if(EvaluableNode::IsEmptyNode(id_path))
-		return std::make_pair(EntityReferenceType(from_entity), ContainerEntityReferenceType(nullptr));
+		//find first index
+		while(curIndex < non_null_size && EvaluableNode::IsNull((*idPathEntries)[curIndex]))
+			curIndex++;
 
-	if(id_path->GetType() != ENT_LIST)
-	{
-		//get the string id, get a reference if returning it
-		if(dest_sid_with_reference == nullptr)
-			sid = EvaluableNode::ToStringIDIfExists(id_path);
+		if(has_destination_id)
+		{
+			lastIdIndex = non_null_size - 1;
+
+			//walk down to find the entity id
+			entityIdIndex = lastIdIndex;
+			while(entityIdIndex > curIndex && EvaluableNode::IsNull((*idPathEntries)[entityIdIndex - 1]))
+				entityIdIndex--;
+		}
 		else
-			sid = *dest_sid_with_reference = EvaluableNode::ToStringIDWithReference(id_path);
+		{
+			//index of the id that will be used for the target entity
+			entityIdIndex = non_null_size - 1;
```

Comment:
should be able to put these two lines from this else block above the if(has_destination_id) block and get rid of lines 80 and 83 and this else altogether